subsys/bluetooth/services/ras/rreq/ras_rreq.c

```diff
@@ -0,0 +1,714 @@
+};
+
+static struct bt_ras_rreq m_rreq_pool[CONFIG_BT_RAS_RREQ_MAX_ACTIVE_CONN];
+
+static struct bt_ras_rreq *ras_rreq_find(struct bt_conn *conn)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(m_rreq_pool); i++) {
+		struct bt_ras_rreq *rreq = &m_rreq_pool[i];
+
+		if (rreq->conn == conn) {
+			return rreq;
+		}
+	}
+
+	return NULL;
+}
+
+static int ras_rreq_alloc(struct bt_conn *conn)
+{
+	struct bt_ras_rreq *rreq = NULL;
+
+	if (ras_rreq_find(conn) != NULL) {
+		return -EALREADY;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(m_rreq_pool); i++) {
+		if (m_rreq_pool[i].conn == NULL) {
+			rreq = &m_rreq_pool[i];
+			break;
+		}
+	}
+
+	if (rreq == NULL) {
+		return -ENOMEM;
+	}
+
+	LOG_DBG("conn %p new rreq %p", (void *)conn, (void *)rreq);
+
+	memset(rreq, 0, sizeof(struct bt_ras_rreq));
+	rreq->conn = bt_conn_ref(conn);
+
+	return 0;
+}
+
+static uint8_t ranging_data_ready_notify_func(struct bt_conn *conn,
+					      struct bt_gatt_subscribe_params *params,
+					      const void *data, uint16_t length)
+{
+
+	if (length != sizeof(uint16_t)) {
+		LOG_DBG("Ranging Data Ready Indication size error");
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	uint16_t ranging_counter = *(uint16_t *)data;
+
+	rreq->rd_ready.cb(conn, ranging_counter);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void data_receive_finished(struct bt_ras_rreq *rreq)
+{
+	int error_code;
+
+	if (!rreq->on_demand_rd.last_segment_received) {
+		LOG_DBG("Ranging data completed with missing segments");
+		rreq->on_demand_rd.error_with_data_receive = true;
+	}
+
+	error_code = rreq->on_demand_rd.error_with_data_receive ? -EINVAL : 0;
+	rreq->on_demand_rd.data_get_in_progress = false;
+
+	rreq->on_demand_rd.cb(rreq->conn, rreq->on_demand_rd.counter_in_progress, error_code);
+}
+
+static uint8_t ranging_data_overwritten_notify_func(struct bt_conn *conn,
+						    struct bt_gatt_subscribe_params *params,
+						    const void *data, uint16_t length)
+{
+
+	if (length != sizeof(uint16_t)) {
+		LOG_DBG("Ranging Data Ready Indication size error");
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	struct bt_ras_rreq *rreq = ras_rreq_find(conn);
+
+	uint16_t ranging_counter = *(uint16_t *)data;
+
+	if (rreq->on_demand_rd.data_get_in_progress &&
+	    rreq->on_demand_rd.counter_in_progress == ranging_counter) {
+		if (rreq->cp.state != BT_RAS_RREQ_CP_STATE_NONE) {
+			LOG_DBG("Overwritten received while writing to RAS-CP, will continue "
+				"waiting for RAS-CP response");
+			return BT_GATT_ITER_CONTINUE;
+		}
+
+		LOG_DBG("Ranging counter %d overwritten whilst receiving", ranging_counter);
+		rreq->on_demand_rd.error_with_data_receive = true;
+		data_receive_finished(rreq);
+
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	rreq->rd_overwritten.cb(conn, ranging_counter);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static int ack_ranging_data(struct bt_ras_rreq *rreq)
+{
+	int err;
+
+	NET_BUF_SIMPLE_DEFINE(ack_ranging_data, RASCP_CMD_OPCODE_LEN + sizeof(uint16_t));
```

Comment:
Can we use a different name here instead of `ack_ranging_data` as it's same with the function.